tests/xrt/enqueue/enqueue2.cpp

```diff
@@ -0,0 +1,453 @@
+
+#include <atomic>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <list>
+#include <thread>
+#include <vector>
+
+#ifdef _WIN32
+# pragma warning ( disable : 4267 )
+#endif
+
+// Kernel specifics
+// void addone (__global ulong8 *in1, __global ulong8* in2, __global ulong8* out, unsigned int elements)
+// addone(in1, in2, out, ELEMENTS)
+// The kernel is compiled with 8 CUs same connectivity.
+static constexpr size_t ELEMENTS = 16;
+static constexpr size_t ARRAY_SIZE = 8;
+static constexpr size_t MAXCUS = 8;
+static size_t compute_units = MAXCUS;
+
+static void
+usage()
+{
+  std::cout << "usage: %s [options] \n\n";
+  std::cout << "  -k <bitstream>\n";
+  std::cout << "  -d <device_index>\n";
+  std::cout << "";
+  std::cout << "  [--jobs <number>]: number of concurrently scheduled jobs\n";
+  std::cout << "  [--cus <number>]: number of cus to use (default: 8) (max: 8)\n";
+  std::cout << "  [--seconds <number>]: number of seconds to run\n";
+  std::cout << "";
+  std::cout << "* Program repeatedly enqueues an event graph for specified number of seconds\n";
+  std::cout << "* Summary prints \"jsz sec jobs\" for use with awk, where jobs is total number \n";
+  std::cout << "* of jobs executed in the specified run time\n";
+}
+
+static std::string
+get_kernel_name(size_t cus)
+{
+  std::string k("addone:{");
+  for (int i=1; i<cus; ++i)
+    k.append("addone_").append(std::to_string(i)).append(",");
+  k.append("addone_").append(std::to_string(cus)).append("}");
+  return k;
+}
+
+// Flag to stop job rescheduling.  Is set to true after
+// specified number of seconds.
+static std::atomic<bool> stop{true};
+
+// Create queues to execute the event graph
+static xrt::queue q0;
+static xrt::queue q1;
+static xrt::queue q2;
+static xrt::queue q3;
+static xrt::queue q4;
+static xrt::queue q5;
+
+// Data for a single job
+struct job_type
+{
+  size_t id = 0;
+  size_t runs = 0;
+
+  // Kernel object
+  // void addone (__global ulong8 *in1, __global ulong8* in2, __global ulong8* out, unsigned int elements)
+  xrt::kernel k;
+
+  // Runs for the job
+  std::array<xrt::run, 7> r;
+
+  // Input buffers for the runs
+  std::array<xrt::bo, 6> a;
+
+  // Output buffers
+  std::array<xrt::bo, 7> o;
+
+  // a events
+  std::array<xrt::queue::event, 6> ea;
+
+  // run events
+  std::array<xrt::queue::event, 7> er;
+
+  // output events
+  std::array<xrt::queue::event, 7> eo;
+
+  job_type(const xrt::device& device, xrt::kernel krnl)
+    : k(std::move(krnl))
+  {
+    static size_t count=0;
+    id = count++;
+
+    auto grpid0 = k.group_id(0);
+    auto grpid1 = k.group_id(1);
+    const size_t data_size = ELEMENTS * ARRAY_SIZE;
+
+    for (int i = 0; i < 7; ++i)
+      r[i] = xrt::run(k);
+
+    for (int i = 0; i < 6; ++i)
+      a[i] = xrt::bo(device, data_size * sizeof(unsigned long), grpid0);
+
+    for (int i = 0; i < 7; ++i)
+      o[i] = xrt::bo(device, data_size * sizeof(unsigned long), grpid1);
+  }
+
+  job_type(job_type&& rhs)
```

Comment:
Do you need a non-default special constructor here?

Reply:
Probably not.